public/app/plugins/datasource/prometheus/querycache/QueryCache.ts

```diff
@@ -0,0 +1,260 @@
+      let targExpr = interpolateString(targ.expr);
+      let targSig = getTargSig(targExpr, request, targ); // ${request.maxDataPoints} ?
+
+      reqTargSigs.set(targIdent, targSig);
+    });
+
+    // figure out if new query range or new target props trigger full cache invalidation & re-query
+    for (const [targIdent, targSig] of reqTargSigs) {
+      let cached = this.cache.get(targIdent);
+      let cachedSig = cached?.sig;
+
+      if (cachedSig !== targSig) {
+        doPartialQuery = false;
+      } else {
+        // only do partial queries when new request range follows prior request range (possibly with overlap)
+        // e.g. now-6h with refresh <= 6h
+        prevTo = cached?.prevTo ?? Infinity;
+        doPartialQuery = newTo > prevTo && newFrom <= prevTo;
+      }
+
+      if (!doPartialQuery) {
+        break;
+      }
+    }
+
+    // console.log(`${doPartialQuery ? 'partial' : 'full'} query`);
+
+    if (doPartialQuery) {
+      // 10m re-query overlap
+
+      // clamp to make sure we don't re-query previous 10m when newFrom is ahead of it (e.g. 5min range, 30s refresh)
+      let newFromPartial = Math.max(prevTo! - this.overlapWindow, newFrom);
+
+      // modify to partial query
+      request = {
+        ...request,
+        range: {
+          ...request.range,
+          from: dateTime(newFromPartial),
+          to: dateTime(newTo),
+        },
+      };
+    } else {
+      reqTargSigs.forEach((targSig, targIdent) => {
+        this.cache.delete(targIdent);
+      });
+    }
+
+    return {
+      requests: [request],
+      targSigs: reqTargSigs,
+      shouldCache,
+    };
+  }
+
+  // should amend existing cache with new frames and return full response
+  procFrames(
+    request: DataQueryRequest<PromQuery>,
+    requestInfo: CacheRequestInfo | undefined,
+    respFrames: DataFrame[]
+  ): DataFrame[] {
+    if (requestInfo?.shouldCache) {
+      const newFrom = request.range.from.valueOf();
+      const newTo = request.range.to.valueOf();
+
+      // group frames by targets
+      const respByTarget = new Map<TargetIdent, DataFrame[]>();
+
+      respFrames.forEach((frame: DataFrame) => {
+        let targIdent = `${request.dashboardUID}|${request.panelId}|${frame.refId}`;
+
+        let frames = respByTarget.get(targIdent);
+
+        if (!frames) {
+          frames = [];
+          respByTarget.set(targIdent, frames);
+        }
+
+        frames.push(frame);
+      });
+
+      let outFrames: DataFrame[] = [];
+
+      respByTarget.forEach((respFrames, targIdent) => {
+        let cachedFrames = (targIdent ? this.cache.get(targIdent)?.frames : null) ?? [];
+
+        respFrames.forEach((respFrame: DataFrame) => {
+          // skip empty frames
+          if (respFrame.length === 0 || respFrame.fields.length === 0) {
+            return;
+          }
+
+          // frames are identified by their second (non-time) field's name + labels
+          // TODO: maybe also frame.meta.type?
```

Comment:
This should also be trackable in an issue. Could you please create one?

Reply:
i'm not sure if this is a real todo, it was just a possible thought in case we didnt have enough determinism here. if everything works okay then we can just remove it, but would also be useful to leave the comment in for a while so it's easy to find exact place to adjust.